src/Type/Php/StrIncrementDecrementFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,136 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\Type;
+use ValueError;
+use function chr;
+use function count;
+use function function_exists;
+use function implode;
+use function in_array;
+use function is_float;
+use function is_int;
+use function is_numeric;
+use function ord;
+use function preg_match;
+use function str_split;
+use function stripos;
+
+class StrIncrementDecrementFunctionReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), ['str_increment', 'str_decrement'], true);
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): Type
+	{
+		$fnName = $functionReflection->getName();
+		$args = $functionCall->getArgs();
+
+		if (count($args) !== 1) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$argType = $scope->getType($args[0]->value);
+		if (count($argType->getConstantScalarValues()) === 0) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$value = $argType->getConstantScalarValues()[0];
```

Comment:
You should account for union of values, don't just read the first one

Reply:
I see, I misunderstood `UnionType::getConstantScalarValues()`.